File: suspension_model/suspension_elements/secondary_elements/steering_link.py
from suspension_model.assets.misc_linalg import rotation_matrix
from suspension_model.suspension_elements.primary_elements.link import Link
from suspension_model.suspension_elements.primary_elements.node import Node
import numpy as np


class SteeringLink(Link):
    def __init__(self, inboard: Node, outboard: Node, kingpin: Link) -> None:
        super().__init__(inboard=inboard, outboard=outboard)

        self.kingpin = kingpin
        self.initial_length = self.length
        self.steering_pickup_to_kingpin = self._steering_pickup_to_kingpin
        self.angle = 0

    @property
    def length(self):
        return np.linalg.norm(self.inboard_node.position - self.outboard_node.position)

    @property
    def _steering_pickup_to_kingpin(self):
        # Steering pickup to kingpin
        ang_x, ang_y = self.kingpin.normalized_transform()
        steering_pickup_pos_shifted = self.outboard_node.position - self.kingpin.inboard_node.position
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=-1 * ang_y)
        
        return np.matmul(y_rot, np.matmul(x_rot, steering_pickup_pos_shifted))

    def update(self):
        # Recalculate steering pickup based on jounce condition
        ang_x, ang_y = self.kingpin.normalized_transform()
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=-1 * ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=ang_y)
        z_rot = rotation_matrix(unit_vec=[0, 0, 1], theta=self.angle)
        steering_pickup_position = np.matmul(y_rot, np.matmul(x_rot, np.matmul(z_rot, self.steering_pickup_to_kingpin))) + self.kingpin.inboard_node.position

        self.outboard_node.position = steering_pickup_position
    
    def rotate(self, angle: float):
        self._set_initial_position()
        self.angle = angle
        self.update()
    
    def _set_initial_position(self):
        kingpin_rot = rotation_matrix(unit_vec=self.kingpin.direction, theta=-1 * self.angle)
        steer_about_origin = self.outboard_node.position - self.kingpin.inboard_node.position
        self.outboard_node.position = np.matmul(kingpin_rot, steer_about_origin) + self.kingpin.inboard_node.position

File: kin_cpp/secondary_elements/steering_link.cpp
#include "steering_link.h"
#include "../assets/misc_linalg.h"

SteeringLink::SteeringLink (Beam *sb, Kingpin *kp) {
    steering_beam = sb;
    kingpin = kp;
    initial_length = length ();
    angle = 0.0;
}

double SteeringLink::length () const {
    return norm (steering_beam->getInboardNode ()->position - steering_beam->getOutboardNode ()->position);
}

StaticVector<double, 3UL> SteeringLink::_steering_pickup_to_kingpin () const {
    StaticVector<double, 2UL> xy_angles = kingpin->getBeam ()->normalized_transform ();
    StaticVector<double, 3UL> steering_pickup_pos_shifted = steering_beam->getOutboardNode ()->position - steering_beam->getInboardNode ()->position;
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix ({1, 0, 0}, xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix ({0, 1, 0}, -1 * xy_angles[1]);

    return y_rot * (x_rot * steering_pickup_pos_shifted);
}
void SteeringLink::update () {
    StaticVector<double, 2UL> xy_angles = kingpin->getBeam ()->normalized_transform ();
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix ({1, 0, 0}, -1 * xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix ({0, 1, 0}, xy_angles[1]);
    StaticMatrix<double, 3UL, 3UL> z_rot = rotation_matrix ({0, 0, 1}, angle);
    StaticVector<double, 3UL> steering_pickup_position = x_rot * (y_rot * (z_rot * _steering_pickup_to_kingpin ())) + kingpin->getBeam ()->getInboardNode ()->position;

    steering_beam->getOutboardNode ()->position = steering_pickup_position;
}

void SteeringLink::rotate (double new_angle) {
    _set_initial_position ();
    angle = new_angle;
    update ();
}
void SteeringLink::_set_initial_position () {
    StaticMatrix<double, 3UL, 3UL> kingpin_rot = rotation_matrix (kingpin->getBeam ()->direction (), -1 * angle);
    StaticVector<double, 3UL> steer_about_origin = steering_beam->getOutboardNode ()->position - kingpin->getBeam ()->getInboardNode ()->position;
    steering_beam->getOutboardNode ()->position = kingpin_rot * steer_about_origin + kingpin->getBeam ()->getInboardNode ()->position;
}